flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java

```diff
@@ -0,0 +1,192 @@
+			if (!stateAndRefCnt.f0.equals(state)) {
+				throw new IllegalStateException("Cannot register a key with different states.");
+			}
+
+			stateAndRefCnt.f1++;
+		}
+	}
+
+	/**
+	 * Decrease the reference count of the state in the registry
+	 *
+	 * @param key The key of the state to unregister
+	 */
+	public void unregister(String key) {
+		Tuple2<StateObject, Integer> stateAndRefCnt = registeredStates.get(key);
+
+		if (stateAndRefCnt == null) {
+			throw new IllegalStateException("Cannot unregister an unexisted state.");
+		}
+
+		stateAndRefCnt.f1--;
+
+		// Remove the state from the registry when it's not referenced any more.
+		if (stateAndRefCnt.f1 == 0) {
+			registeredStates.remove(key);
+			discardedStates.add(stateAndRefCnt.f0);
+		}
+	}
+
+	/**
+	 * Register all the shared states in the given state handles.
+	 * 
+	 * @param stateHandles The state handles to register their shared states
+	 */
+	public void registerAll(Collection<? extends CompositeStateHandle> stateHandles) {
+		synchronized (this) {
+			if (stateHandles != null) {
+				for (CompositeStateHandle stateHandle : stateHandles) {
+					stateHandle.register(this);
+				}
+			}
+		}
+	}
+	
+	/**
+	 * Register all the shared states in the given state handle.
+	 * 
+	 * @param stateHandle The state handle to register its shared states
+	 */
+	public void registerAll(CompositeStateHandle stateHandle) {
+		if (stateHandle != null) {
+			synchronized (this) {
+				stateHandle.register(this);
+			}
+		}
+	}
+
+	/**
+	 * Unregister all the shared states in the given state handles and return
+	 * those unreferenced states after these shared states are unregistered.
+	 * 
+	 * @param stateHandles The state handles to unregister their shared states
+	 * @return The states that are not referenced any more
+	 */
+	public List<StateObject> unregisterAll(Collection<? extends CompositeStateHandle> stateHandles) {
+		synchronized (this) {
+			discardedStates.clear();
+
+			if (stateHandles != null) {
+				for (CompositeStateHandle stateHandle : stateHandles) {
+					stateHandle.unregister(this);
+				}
+			}
+
+			return discardedStates;
+		}
+
+	}
+
+	/**
+	 * Unregister all the shared states in the given state handles and return
+	 * those unreferenced states after these shared states are unregistered.
+	 *
+	 * @param stateHandle The state handle to unregister its shared states
+	 * @return The states that are not referenced any more
+	 */
+	public List<StateObject> unregisterAll(CompositeStateHandle stateHandle) {
+		if (stateHandle == null) {
+			return null;
+		} else {
+			synchronized (this) {
+				discardedStates.clear();
+
+				stateHandle.unregister(this);
+
+				return discardedStates;
+			}
+		}
+	}
+
+	@VisibleForTesting
+	int getReferenceCount(String key) {
+		Tuple2<StateObject, Integer> stateAndRefCnt = registeredStates.get(key);
+
+		return stateAndRefCnt == null ? 0 : stateAndRefCnt.f1;
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) {
```

Comment:
I think `equals()` and `hashCode()` are not really well defined on the state registry, and not needed. I would remove them.